providers/base/bin/prime_offload_tester.py

```diff
@@ -0,0 +1,317 @@
+        if card_name is None:
+            self.logger.info("Couldn't find card name,"
+                             " please check your pci name")
+            return PrimeOffloaderError.NO_CARD_NAME
+
+        # run offload command in other process
+        dri_pci_name_format = re.sub("[:.]", "_", pci_name)
+
+        if timeout > 0:
+            tmp_cmd = "timeout {} {}".format(timeout, cmd)
+        else:
+            tmp_cmd = cmd
+
+        if driver in ('nvidia', 'pcieport'):
+            offload_cmd = "__NV_PRIME_RENDER_OFFLOAD=1" \
+                          " __GLX_VENDOR_LIBRARY_NAME=nvidia {}" \
+                          .format(tmp_cmd)
+        else:
+            offload_cmd = "DRI_PRIME=pci-{} {}" \
+                          .format(dri_pci_name_format, tmp_cmd)
+
+        # if nv driver under nvidia mode, prime/reverse prime couldn't work.
+        if self.check_nv_offload_env() \
+           == PrimeOffloaderError.NOT_SUPPORT_NV_PRIME:
+            self.logger.info("Couldn't use nv prime offload"
+                             " on this system environment")
+            offload_cmd = tmp_cmd
+            # Couldn't work, always return NO_ERROR
+            self.check_result = PrimeOffloaderError.NO_ERROR
+        else:
+            # use other thread to check offload is correctly or not
+            check_thread = threading.Thread(target=self.check_offload,
+                                            args=(cmd, card_id, card_name))
+            check_thread.start()
+
+        offload = subprocess.Popen(offload_cmd,
+                                   shell=True,
+                                   stdout=subprocess.PIPE,
+                                   universal_newlines=True)
+
+        self.logger.info("offload command:[{}]".format(offload_cmd))
+
+        # redirect offload command output real time
+        while offload.poll() is None:
+            line = offload.stdout.readline().strip()
+            self.logger.info(line)
+        return PrimeOffloaderError.NO_ERROR
+
+    def main(self) -> int:
+        """
+        main function for command line processing
+        """
+        parser = argparse.ArgumentParser(
+            prog="Prime offload tester",
+            description="Test prime offload feature",
+        )
+
+        parser.add_argument(
+            "-c", "--command", type=str, default='glxgears',
+            help='command to offload to specific GPU (default: %(default)s)'
+        )
+        parser.add_argument(
+            "-p", "--pci", type=str, default='0000:00:02.0',
+            help='pci name in NNNN:NN:NN.N format (default: %(default)s)'
+        )
+        parser.add_argument(
+            "-d", "--driver", type=str, default='i915',
+            help='Type of GPU driver (default: %(default)s)'
+        )
+        parser.add_argument(
+            "-t", "--timeout", type=int, default=0,
+            help='executing command duration in second (default: %(default)s)'
```

Comment:
From the code, it looks like 0 means no timeout. Maybe it should be explained in the help text.